File: SearchInRotatedSortedArray.cpp
#include <iostream>
using namespace std;

class Solution 
{
public:
    int search(int A[], int n, int target) 
    {
    	if (n == 0)
    		return -1;
    	int l = 0;
    	int r = n - 1;
    	int m;
    	while (l <= r)
    	{
    		m = (l + r) / 2;
    		if (A[m] == target)
    			return m;
    		else if (A[m] < A[r])
    		{
    			if (target >= A[m] && target < A[r])
    				l = m + 1;
    			else
    				r = m - 1;
    		}
    		else
    		{
    			if (target >= A[l] && target < A[m])
    				r = m - 1;
    			else 
    				l = m + 1;
    		}
    	}
    	return -1;
    }
};

int main()
{
	return 0;
}

File: JavaSolution/src/FaceBook/SearchinRotatedSortedArray.java
package FaceBook;

/**
 * Created by xyunpeng on 2/3/16.
 */
public class SearchinRotatedSortedArray {
    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return  -1;
        }

        int start = 0;
        int end = nums.length - 1;
        int mid;
        while (start + 1 < end) {
            mid = start + ((end - start) >> 1);
            if (nums[mid] == target) {
                return mid;
            }

            if (nums[mid] > nums[start]) {
                if (target >= nums[start] && target < nums[mid]) {       //这里注意要加上等于
                    end = mid;
                } else {
                    start = mid;
                }
            } else {
                if (target <= nums[end] && target > nums[mid]) {         //这里注意要加上等于
                    start = mid;
                } else {
                    end = mid;
                }
            }
        }

        if (nums[start] == target) {
            return start;
        }

        if (nums[end] == target) {
            return end;
        }

        return -1;
    }
}
